### DISCUSSION TITLE
Why does reactivity not work when destructure properties from createStore value?

### DISCUSSION THREAD

**Discussion by juniorforlife** [solidjs/solid#1708] [Q&A] [answered]
I have been banging my head on the desk not understanding why reactivity doesn't work with this code:

```
const [authState, setAuthState] = createStore({
  user: null,
  error: null,
  loading: true,
});

const {user} = authState;

return <div>{user?.name}</div>
```
Turns out it only works when I access the properties directly without destructuring
```
const [authState, setAuthState] = createStore({
  user: null,
  error: null,
  loading: true,
});

return <div>{authState.user?.name}</div>
```

Why is this?

**Comment by atk** [comment]
Because stores use a proxy to track access to provide fine-grained reactivity. Once you destructure like in your example, the only access happens outside of an effect or computation, which are the scopes in which the reactive tracking works.

**Reply by juniorforlife** [reply]
sorry I don't understand what you mean. I guess this is a bit low-level so I would appreciate it if you could recommend an in-depth link/blog from Solid team. Thanks

**Reply by atk** [reply, marked as answer]
My apologies, please let me rephrase that in simpler terms.

Solid's reactivity is based on signals. Signals consist of two functions, one to read and one to write their values. To be reactive, the signals require their read function to be called inside another function, referred to as "reactive root" or "tracking scope".

This function keeps track of all read operations made inside certain other functions, so-called "computations", which includes `createMemo`, `createEffect`, everything in squiggly brackets in JSX, and similar functions. While these run inside a root, the root collects all signals that are read by them. Whenever any of those signals change through the setter, the computations are called again.

Signals are great for simple values like numbers and strings, but don't work well with arrays and objects, since those only change if you create new ones; also, that means that a change inside the object would trigger every computation, even if the part that changed was not used by it. To handle those cases, Solid.js provides stores.

Stores are using Proxies. Those are a feature of JavaScript that allows to intercept the access to an object. In this case, the read operation of the store is intercepted to perform a read of a signal for simple values or return another similar proxy for objects and arrays.

That means that whenever you access a simple value from a store, a signal is called under the hood. Unfortunately, if you destructure a store, all subsequent access will be registered from the point of the destructuring, which in your case was outside of the effect, thus the root could not register the underlying signal to re-run anything in case of changes.

I hope that explanation will clarify my answer.

**Reply by juniorforlife** [reply]
great explanation 🙏 thank you

**Comment by snnsnn** [comment]
```js
const {user} = authState
```
Because you are extracting a reactive value into static one and using it.
